File: PAP2223/src/tp01/MyMath.java
package tp01;

/**
 * 
 * @author Didier VO
 *
 */
public class MyMath {

	/**
	 * Vérifie si le nombre est pair
	 * 
	 * @param n entier >=0
	 * @return true si pair
	 */
	public static boolean estPair(int n) {
		assert n >= 0 : "n doit être un entier >=0";
		return n % 2 == 0;
	}

	/**
	 * Calcule le Nème nombre de Fibonacci
	 * 
	 * @param n >=0
	 * @return Fn
	 */
	public static int fibo(int n) {
		assert n >= 0 : "Le nombre doit être >=0";
		int f0 = 0;
		int f1 = 1;
		int f2 = n;
		for (int i = 2; i <= n; i++) {
			f2 = f1 + f0;
			f0 = f1;
			f1 = f2;
		}
		return f2;
	}

	public static DataPhi calculPhi(double prec) {
		assert prec > 0 && prec <= 1 : " La précision doit être <0 et <=1";
		double phi = (1 + Math.sqrt(5)) / 2;
		double f0 = 1;
		double f1 = 1;
		double f2;
		int n = 2;
		while (Math.abs(phi - f1 / f0) > prec) {
			f2 = f1 + f0;
			f0 = f1;
			f1 = f2;
			n++;
		}
		return new DataPhi(n, f1 / f0);
	}

	/**
	 * Calcule le nombre de bits à 1 dans n
	 * 
	 * @param n
	 * @return nombre de bits à 1 dans n
	 */
	public static byte nbBitsV1Bis(int n) {
		int m = 1; // masque 1
		int w = -1; // masque 2
		byte cpt = 0;
		while ((n & w) != 0) {
			if ((n & m) != 0)
				cpt++;
			m = m << 1;
			w = w << 1;
		}
		return cpt;
	}

	/**
	 * Calcule le nombre de bits à 1 dans n
	 * 
	 * @param n
	 * @return nombre de bits à 1 dans n
	 */
	public static byte nbBitsV2Bis(int n) {
		byte cpt = 0;
		while (n != 0) {
			cpt += n & 1;
			n = n >>> 1;
		}
		return cpt;
	}

	/**
	 * Calcule la distance d'un projectile
	 * 
	 * @param v m/s
	 * @param a angle entre 1 et 89n degré
	 * @return
	 */
	public static double calculDist(double v, double a) {
		assert v > 0 && a >= 1 && a <= 89 : "V doit être >0 et 1<= a <=89";
		a = Math.toRadians(a);
		return 2.0 * v * v * Math.sin(a) * Math.cos(a) / 9.8;
	}

	public static byte getNbBytesUtf8(char c) {
		short i = (short) c;
		if ((i & 0xFF80) == 0)
			return 1;
		if ((i & 0xF800) == 0)
			return 2;
		return 3;
	}

	public static void main(String[] args) {
		DataPhi d = calculPhi(0.000001);
		System.out.println(d.n() + " : " + d.phi());

		/*
		 * System.out.println("Entrez un nombre: "); Scanner console = new
		 * Scanner(System.in); int n = console.nextInt(); System.out.println(n +
		 * " est pair :" + estPair(n)); console.close();
		 */
	}
}

File: PAP2223/python/mymath.py
import math

def estPair(n : int)-> bool:
    '''indique si un nombre est pair'''
    assert type(n) == int and n>=0, "Le nombre doit être un entier >=0"
    return n%2==0

def fibo(n : int) -> int :
    '''Calcule le nème nbr de Fibonacci '''
    assert type(n)==int and n>=0 , " Doit être un entier >=0"
    f0 , f1 , f2 = 0 , 1 , n
    for i in range(1,n):
        f2 = f1 + f0
        f1 , f0 = f2 , f1
    return f2

def calculPhi(prec : float)-> (int,float):
    '''Calcul le nombre d'itérations pour avoir le nombre d'or
        avec une précision donnée '''
    assert 0< prec <=1 , "Doit respecter: 0< prec <=1"
    phi= (1 + math.sqrt(5))/2
    f0 = f1 = 1
    n = 2
    while math.fabs(phi - f1 / f0)> prec:
        f2 = f1 + f0
        f0 , f1 = f1 , f2
        n = n + 1
    return n , f1/f0


def nbBitsV1(n:int)->int:
    '''Calcule le nombre de bits à 1 au minimum pour coder le nombre n'''
    taille=n.bit_length()
    # comptabilise le bit de signe si négatif
    cpt = 0 if n>=0 else 1
    m = 1
    for i in range(0,taille):
        if (n & m)!=0 :
            cpt = cpt + 1
        m = m << 1
    return  cpt
            
    
    

    
def nbBitsV2Bis(n:int)->int:
    '''Calcule le nbr de bits à 1 sur une zone de 32bits'''
    assert  -(1<<31)<= n < (1<<31) , "le nombre doit tenir sur 32bits"
    #Si négatif conversion en un motif sur 32 bits entier
    if n<0 :
        n = (1<< 32) + n
    cpt = 0
    while n != 0 :
        cpt = cpt + (n & 1)
        n= n>>1
    return cpt

    
    
    
    

#print("Nom module: ",__name__)

if __name__ == "__main__":
    print(nbBitsV2Bis(-1))
    print(f'Calcul n: {calculPhi(0.1)}')
    print(f'Le nombre 20 est pair {estPair(20)}')
    print(f'Le nombre 27 est pair {estPair(27)}')